### PR TITLE
out_kafka: add record-derived Kafka message headers

### LINKED ISSUE #12286: out_kafka: support record-derived Kafka message headers
**Is your feature request related to a problem? Please describe.**

The Kafka protocol and Fluent Bit's bundled librdkafka client support message headers, but the `out_kafka` plugin doesn't provide a general way to create those headers from a record field.

Some pipelines receive a variable set of per-record transport attributes, such as correlation identifiers, routing metadata, or content metadata. These values need to be Kafka message headers rather than fields in the serialized message body. Keeping them in the payload isn't equivalent because Kafka consumers, interceptors, and routing infrastructure access message headers separately from the value.

librdkafka already exposes the required producer APIs through `rd_kafka_headers_new()`, `rd_kafka_header_add()`, and `RD_KAFKA_V_HEADERS`. The missing part is the Fluent Bit `out_kafka` configuration and record-mapping layer.

**Describe the solution you'd like**

Add a `headers_key` option to `out_kafka`. Its value identifies a top-level record field whose MessagePack map is converted to Kafka message headers for that record.

Proposed behavior:

- An empty or ASCII-whitespace-only `headers_key` disables the feature.
- Nonblank field names are matched exactly and aren't trimmed.
- The selected field must be a MessagePack map.
- Header names must be strings.
- Header values can be strings, binary values, or null values.
- Empty values, null values, insertion order, and duplicate names are preserved on the Kafka record.
- Unsupported map entries are skipped with a warning.
- A valid selected map is removed from the serialized payload by default.
- A `preserve_headers_key` boolean option, defaulting to `false`, retains the original map in the payload when enabled.
- A missing selected field leaves payload processing unchanged.
- A selected field with a non-map value remains in the payload to avoid data loss.
- An empty map is valid and follows the configured preservation behavior.
- If duplicate top-level map fields match, the first map supplies the Kafka headers and later fields remain in the payload.
- The feature applies to the `avro`, `gelf`, `json`, `msgpack`, and `raw` formats. A nonblank `headers_key` is rejected for `otlp_json` and `otlp_proto`, whose serialization path doesn't expose an individual source map.
- Kafka 0.11-compatible brokers or later are required for message-header support.

Sanitized YAML example:

```yaml
pipeline:
  inputs:
    - name: dummy
      tag: example.logs
      dummy: '{"message":"example","kafka_headers":{"trace-id":"abc-123","content-type":"application/json"}}'

  outputs:
    - name: kafka
      match: example.logs
      brokers: 192.0.2.10:9092
      topics: example-logs
      format: json
      headers_key: kafka_headers
      preserve_headers_key: false
```

This produces `trace-id` and `content-type` as Kafka message headers. The `kafka_headers` field isn't included in the JSON value unless `preserve_headers_key` is set to `true`.

**Describe alternatives you've considered**

- Keep the attributes in the serialized payload. This doesn't make them available through Kafka's message-header interface.
- Configure individual headers explicitly. PR #8583 proposes repeated `header` options with static values or individual record-field references. That interface is useful when the header names are known in configuration, but it doesn't cover records that already contain a variable header map. The two interfaces are complementary.
- Transform the record with a filter or Lua script. Filters can reshape the payload, but they can't attach Kafka producer headers through the existing output interface.
- Patch librdkafka. This isn't necessary because librdkafka already implements the required protocol and ownership semantics.

**Additional context**

This request is a focused `out_kafka` implementation of the record-map use case discussed in #1673, #8334, and #9448. Those broader requests are closed as stale; #8334 and #9448 also cover both Kafka input and output, while this request deliberately defines and tests the output-side contract. It is intended to coexist with the explicit per-header configuration proposed in #8583:

- `header`: configure one known header at a time, optionally sourcing its value from a record field.
- `headers_key`: convert all supported entries from one record map at runtime.

The implementation can remain confined to Fluent Bit-owned `plugins/out_kafka` code and tests. It doesn't require changes to bundled librdkafka sources or to the Kafka wire protocol.

A tested implementation and separate documentation change are prepared locally. Coverage includes payload removal and preservation, empty/missing/non-map sources, null and binary values, duplicate names, MessagePack, raw, GELF, Avro with Schema Registry resolution, queue-full retry ownership, configured message-size rejection, startup validation, and strict Valgrind execution.

Related work: #1673, #8334, and #9448 (closed requests); #8583 and #9057 (open, conflicting implementations of individually configured headers).

### PR DESCRIPTION
**Pull request by kimonus**
## Summary

Add record-derived Kafka message headers to the `out_kafka` plugin.

`headers_key` selects a top-level MessagePack map in each log record. Supported
entries are transferred to librdkafka as Kafka message headers. The selected
map is omitted from the serialized value by default; `preserve_headers_key
true` retains it.

The implementation preserves string, binary, empty, and null values, map
order, and duplicate header names. Invalid entries are skipped with warnings.
Missing and non-map source fields don't cause payload data loss.

The change uses librdkafka's existing header APIs and
`rd_kafka_producev()`. It keeps header ownership correct across successful
enqueue, enqueue failure, and the existing queue-full retry loop. No bundled
code under `lib/` is modified.

Closes #12286.

This complements #8583: its repeated `header` option configures known headers
individually, while `headers_key` converts a variable map already present in
each record.

## Rebase and Protobuf integration

This branch is rebased onto current `master` after #12345 and #12411. The
`out_kafka` conflicts were resolved while preserving the newer Protobuf and
Schema Registry implementation.

Record-derived headers now have explicit Protobuf coverage. The selected
header map is removed before JSON-to-Protobuf conversion, while the extracted
values are attached to the Kafka record. Protobuf payload cleanup is also
handled if Kafka header construction fails.

## Compatibility

- Supported formats: `avro`, `gelf`, `json`, `msgpack`, `protobuf`, and `raw`.
- Unsupported formats: `otlp_json` and `otlp_proto`; a nonblank `headers_key`
  is rejected during initialization.
- Empty and ASCII-whitespace-only `headers_key` values disable the feature.
- Kafka 0.11-compatible brokers or later are required for message headers.
- Existing behavior is unchanged when `headers_key` isn't configured.

## Before and after

Given this record:

```json
{
  "message": "example",
  "kafka_headers": {
    "trace-id": "abc-123",
    "content-type": "application/json"
  }
}
```

and this output configuration:

```yaml
outputs:
  - name: kafka
    match: example.logs
    brokers: 192.0.2.10:9092
    topics: example-logs
    format: json
    headers_key: kafka_headers
```

before this change, `kafka_headers` can only remain in the serialized value.
After this change, its entries are Kafka message headers and the JSON value is
`{"@timestamp":...,"message":"example"}`. Set
`preserve_headers_key: true` to keep the original map in the value as well.

## Testing

The branch was configured and fully rebuilt in a clean Ubuntu 24.04 container:

```bash
cmake -S . -B build \
  -DFLB_TESTS_RUNTIME=On \
  -DFLB_TESTS_INTERNAL=On \
  -DFLB_AVRO_ENCODER=On \
  -DFLB_PROTOBUF_ENCODER=On \
  -DCMAKE_BUILD_TYPE=Debug
cmake --build build -j8
```

Focused CTest coverage:

```bash
ctest --test-dir build \
  -R 'flb-rt-out_kafka|flb-it-(kafka_schema_registry|kafka_protobuf|avro)' \
  --output-on-failure
```

Result: 4/4 passed.

Complete focused integration coverage:

```bash
FLUENT_BIT_BINARY=/workspace/build/bin/fluent-bit \
  tests/integration/.venv/bin/python -m pytest \
  tests/integration/scenarios/out_kafka/tests/test_out_kafka_001.py \
  tests/integration/scenarios/out_kafka/tests/test_out_kafka_schema_registry.py \
  -q
```

Result: 88 passed.

Strict Valgrind coverage:

```bash
VALGRIND=1 VALGRIND_STRICT=1 \
  FLUENT_BIT_BINARY=/workspace/build/bin/fluent-bit \
  tests/integration/.venv/bin/python -m pytest \
  tests/integration/scenarios/out_kafka/tests/test_out_kafka_001.py \
  tests/integration/scenarios/out_kafka/tests/test_out_kafka_schema_registry.py \
  -q
```

Result: 88 passed with strict Valgrind enabled.

The full enabled CTest inventory was also exercised. All 215 tests other than
`flb-rt-out_td` passed. The sole failure is outside this PR and unchanged from
`upstream/master`: `tests/runtime/out_td.c` creates an unconfigured Treasure
Data output before loading `/tmp/td.conf`, so initialization fails before the
test can run. Initial runtime-shell failures caused by missing `curl` and `nc`
in the clean image were rerun successfully after installing those tools.

Additional checks:

```text
git diff --check: passed
PR-range commit-prefix validation: passed
DCO sign-offs: present on all 11 commits
```

Packaging is not affected, so the packaging-specific checklist is not
applicable.

## Documentation

- [x] Documentation required for this feature

Documentation PR: `fluent/fluent-bit-docs#2672`

The docs PR will be rebased after `fluent/fluent-bit-docs#2717` merges, as
requested by the docs maintainer.

## Backporting

- [ ] Backport to latest stable release.

No backport is requested. This feature targets `master` for the next release.

Fluent Bit is licensed under Apache 2.0. By submitting this pull request, I
understand that this code will be released under the terms of that license.

## Summary by CodeRabbit

* **New Features**
  * Kafka output supports dynamic message headers sourced from record maps.
  * Supports string, binary, null, empty, duplicate, and batched header values.
  * Added options to select the header field and preserve it in the payload.
  * Supports dynamic headers across supported JSON, MessagePack, raw, GELF, and Avro formats.

* **Bug Fixes**
  * Invalid, oversized, blank, and unsupported header configurations are rejected safely.

* **Tests**
  * Expanded coverage for retries, batching, schema-based formats, and edge cases.

### CODE REVIEW

**Comment by kimonus** [comment]
## Test evidence

### Example configuration

```yaml
pipeline:
  inputs:
    - name: dummy
      tag: example.logs
      dummy: '{"message":"example","kafka_headers":{"trace-id":"abc-123","content-type":"application/json"}}'

  outputs:
    - name: kafka
      match: example.logs
      brokers: 192.0.2.10:9092
      topics: example-logs
      format: json
      headers_key: kafka_headers
      preserve_headers_key: false
```

### Functional integration

```bash
FLUENT_BIT_BINARY=/tmp/fluent-bit \
  tests/integration/.venv/bin/python -m pytest \
  tests/integration/scenarios/out_kafka/tests/test_out_kafka_001.py \
  -q -k dynamic_headers
```

```text
25 passed, 29 deselected
```

The complete current `out_kafka` integration scenario also passed:

```text
54 passed
```

The focused cases validate Kafka magic-v2 wire output, exact header bytes,
empty versus null values, binary values, order, duplicate names, payload
removal and preservation, MessagePack, raw, GELF, Avro, batching, queue-full
retry ownership, message-size rejection, and startup validation.

### Strict Valgrind

```bash
VALGRIND=1 VALGRIND_STRICT=1 FLUENT_BIT_BINARY=/tmp/fluent-bit \
  tests/integration/.venv/bin/python -m pytest \
  tests/integration/scenarios/out_kafka/tests/test_out_kafka_001.py \
  -q -k dynamic_headers
```

```text
25 passed, 29 deselected
```

The strict integration harness completed successfully with no Valgrind memory
errors or leaks.

### Runtime and broker smoke tests

`flb-rt-out_kafka` passed all four configuration cases: raw format, an empty
selector, an ASCII-whitespace-only selector, and exact matching for a nonblank
selector.

A Redpanda v24.3.11 smoke test confirmed that the broker received the expected
`trace-id`, empty, and null headers. With `preserve_headers_key` disabled, the
source map was absent from the JSON payload. With preservation enabled, the
same headers were attached and the original source map remained in the JSON
payload.

The build used these options and completed all source and test targets:

```text
-DFLB_TESTS_RUNTIME=On
-DFLB_TESTS_INTERNAL=On
-DFLB_AVRO_ENCODER=On
-DCMAKE_BUILD_TYPE=Debug
```

**Comment by kimonus** [comment]
The initial CentOS 7 compile check exposed a test-registration mismatch: `tests/runtime/out_kafka.c` was registered when `FLB_KAFKA=Off`. Because the new runtime test includes `kafka_config.h`, that configuration attempted to include the unavailable `rdkafka.h` header.

Commit `05989ae61` gates the Kafka runtime test on `FLB_KAFKA`, matching the Kafka plugin registration. Kafka-enabled builds still register the test through the existing `FLB_OUT_KAFKA` argument.

Verification:

- Reproduced the CI configuration with `docker build -t fluent-bit-pr12287-centos7:local -f dockerfiles/Dockerfile.centos7 .`.
- The complete CentOS 7 image build passed.
- The prior `rdkafka.h: No such file or directory` failure is no longer reached.
- No feature implementation code changed; the Kafka-enabled runtime, integration, strict Valgrind, and Redpanda results reported above remain applicable.

File: plugins/out_kafka/kafka.c [hunk 1]
```diff
@@ -1607,6 +1811,23 @@ static struct flb_config_map config_map[] = {
...
+    "option. Nonblank values are matched exactly. This option is supported for "
+    "standard log formats, but not OTLP formats."
+   },
+   {
+    FLB_CONFIG_MAP_BOOL, "preserve_headers_key", "false",
+    0, FLB_TRUE, offsetof(struct flb_out_kafka, preserve_headers_key),
+    "Keep the map selected by headers_key in the record passed to the normal "
+    "payload formatter. By default, a valid selected map is omitted from the "
+    "payload after its entries are converted to Kafka message headers."
+   },
```
[the remark this answers is not in the document]
Comment (kimonus) [reply plugins/out_kafka/kafka.c]: Fixed in `8eb413716` and `091064ad0`. The new Kafka config-map initializers and every `TEST_LIST` entry now use four-space indentation. `git diff --check` and the full PR-range commit-prefix check pass, and `flb-rt-out_kafka` passes all four cases.

File: tests/integration/scenarios/out_kafka/tests/test_out_kafka_001.py [hunk 2]
```diff
@@ -590,6 +728,486 @@ def test_out_kafka_message_key_field_sets_kafka_key():
...
+
+    messages = service.wait_for_messages(1)
+    log_text = _wait_for_log_text(
+        service.service.flb.log_file,
+        "skipping Kafka header with a non-string name",
+    )
+    log_text = _wait_for_log_text(
+        service.service.flb.log_file,
+        "value must be a string, binary, or null",
+    )
```
[the remark this answers is not in the document]
Comment (kimonus) [reply tests/integration/scenarios/out_kafka/tests/test_out_kafka_001.py]: Fixed in `dc417fd5b`. Both tests now retain the non-string-name and invalid-value log results separately and assert both warning texts. The focused scenario passes normally (`25 passed, 29 deselected`) and with strict Valgrind (`25 passed, 29 deselected`, no memory errors or leaks).

**Review by cosmo0920 — DISMISSED** [review]
On my side, it looks good.

**Comment by kimonus** [comment]
@cosmo0920 Could you please rerun the failed Ubuntu unit-test job?

The only leaf failure was `flb-rt-out_s3` in the `-DFLB_SMALL=On` Clang matrix: `compression_snappy_putobject` was interrupted by `SIGSEGV` after reporting a successful object upload. That job passed 201 of 202 tests. The failure is outside the changed `out_kafka` component; the Kafka-focused integration and strict Valgrind checks, sanitizers, other Ubuntu matrices, Windows checks, compile checks, lint, and DCO all passed.

I have not changed or retriggered the branch because there is no evidence that the Kafka patch caused this S3 runtime-test failure.

**Comment by kimonus** [comment]
@edsiper When convenient, could you please take a look at this PR?

@cosmo0920 has approved it, all review threads are resolved, and the required checks have passed. The only non-green leaf job is a non-required macOS ThreadSanitizer run that was cancelled at the one-hour limit; the required aggregate unit-test check passed.

Do you think this feature would be suitable for v5.1.2, or should it target a later release? I’m happy to update the branch and rerun the checks if you would prefer that before merging. Thank you.

**Comment by kimonus** [comment]
@edsiper @cosmo0920 Thank you for the earlier review. I have rebased this PR
onto current `master` and resolved the `out_kafka` conflicts introduced by the
recent Kafka work, including #12411.

The integration preserves the new Protobuf/Schema Registry flow and now
explicitly tests record-derived Kafka headers with a Protobuf payload. I also
fixed cleanup of the Protobuf payload if Kafka header construction fails.

Verification after the rebase:

- clean Ubuntu 24.04 configure and full build passed;
- focused runtime/internal CTest: 4/4 passed;
- complete focused Kafka integration: 88 passed;
- the same 88 tests passed with strict Valgrind enabled;
- 215/216 tests in the full enabled CTest inventory passed. The sole failure,
  `flb-rt-out_td`, is unchanged from `master` and fails during its unrelated
  test setup before Kafka code is involved;
- full PR-range commit-prefix validation, DCO, and `git diff --check` passed.

The branch is current and no longer has the prior merge conflicts. When
convenient, could you please re-review it for inclusion in the next release?
Thank you.

**Comment by kimonus** [comment]
@cosmo0920 @edsiper The rebased branch is green except for one unrelated failure in the Clang MemorySanitizer job. `flb-rt-out_s3` failed because `ordered_construct_error_exhausts_chunk` was interrupted by `SIGSEGV`; the other 207 CTest targets passed. This PR does not modify the S3 output or its tests. Could you please rerun that job when convenient?

The rebase also dismissed the previous approval. Before we continue rebasing this work around further Kafka changes, could you confirm whether record-derived Kafka headers still fit the intended direction for `out_kafka` and whether you would like this PR to proceed toward merge? If so, I will keep the branch current and address any further review feedback.
